server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java

```diff
@@ -212,6 +221,111 @@ public void testDeleteCurrentWriteIndexForDataStream() {
+        metadataBuilder.putCustom(RepositoriesMetadata.TYPE, new RepositoriesMetadata(List.of(repositoryMetadata)));
+        final RoutingTable.Builder routingBuilder = RoutingTable.builder();
+
+        final SnapshotId snapshotId = new SnapshotId(randomAlphaOfLength(10), UUIDs.randomBase64UUID());
+        final Set<Index> indices = new HashSet<>();
+
+        final int nbIndices = randomIntBetween(2, 10);
+        for (int i = 0; i < nbIndices; i++) {
+            Settings.Builder indexSettingsBuilder = Settings.builder()
+                .put("index.version.created", VersionUtils.randomVersion(random()))
+                .put(IndexModule.INDEX_STORE_TYPE_SETTING.getKey(), SearchableSnapshotsSettings.SEARCHABLE_SNAPSHOT_STORE_TYPE)
+                .put(SearchableSnapshotsSettings.SEARCHABLE_SNAPSHOTS_DELETE_SNAPSHOT_ON_INDEX_DELETION, true)
+                .put(SEARCHABLE_SNAPSHOTS_REPOSITORY_NAME_SETTING_KEY, repositoryMetadata.name())
+                .put(SEARCHABLE_SNAPSHOTS_SNAPSHOT_NAME_SETTING_KEY, snapshotId.getName())
+                .put(SEARCHABLE_SNAPSHOTS_SNAPSHOT_UUID_SETTING_KEY, snapshotId.getUUID());
+            if (randomBoolean()) {
+                indexSettingsBuilder.put(SEARCHABLE_SNAPSHOTS_REPOSITORY_UUID_SETTING_KEY, repositoryMetadata.uuid());
+            }
+            IndexMetadata indexMetadata = IndexMetadata.builder(randomAlphaOfLength(10) + i)
+                .settings(indexSettingsBuilder.build())
+                .numberOfShards(randomIntBetween(1, 3))
+                .numberOfReplicas(randomInt(1))
+                .build();
+            metadataBuilder.put(indexMetadata, false);
+            routingBuilder.addAsNew(indexMetadata);
+            indices.add(indexMetadata.getIndex());
+        }
+
+        ClusterState clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .routingTable(routingBuilder.build())
+            .metadata(metadataBuilder)
+            .build();
+
+        SnapshotDeletionsPending pendingDeletions = clusterState.custom(SnapshotDeletionsPending.TYPE, SnapshotDeletionsPending.EMPTY);
+        while (indices.size() > 0) {
+            assertThat(pendingDeletions.isEmpty(), equalTo(true));
+
+            List<Index> indicesToDelete = randomSubsetOf(randomIntBetween(1, Math.max(1, indices.size() - 1)), indices);
+            clusterState = service.deleteIndices(clusterState, Set.copyOf(indicesToDelete));
+            indicesToDelete.forEach(indices::remove);
+
+            for (Index deletedIndex : indicesToDelete) {
+                assertThat(clusterState.metadata().index(deletedIndex), nullValue());
+                assertThat(clusterState.routingTable().index(deletedIndex), nullValue());
+            }
+
+            pendingDeletions = clusterState.custom(SnapshotDeletionsPending.TYPE, SnapshotDeletionsPending.EMPTY);
+        }
+
+        assertThat(pendingDeletions.isEmpty(), equalTo(false));
```

Comment:
Can we also validate the it has size 1, i.e., only led to one pending deletion?

Reply:
Also added in fce436543320a8b2e610280cb8328223e49c3fbd